### PR TITLE
OPT: add Z and SetZ methods

### PR DESCRIPTION
**Pull request by miekg**
Fixes: #1169

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

**Comment by miekg** [comment]
please double check my bit mask, see https://tools.ietf.org/html/rfc6891#section-6.1.3

File: edns.go [hunk 1]
```diff
@@ -148,6 +148,16 @@ func (rr *OPT) SetDo(do ...bool) {
...
 }
 
+// Z returns the Z part of the OPT RR as a uint16 with only the 15 least significant bits used.
+func (rr *OPT) Z() uint16 {
+	return uint16(rr.Hdr.Ttl & 0x007FFF)
+}
+
+// SetZ sets the Z part of the OPT RR, note only the 15 least significant bits of z are used.
+func (rr *OPT) SetZ(z uint16) {
+	rr.Hdr.Ttl = rr.Hdr.Ttl&0xFFFF7FFF | uint32(z)
```
Comment (tmthrgd) [inline edns.go:158]: Shouldn’t this just be `&^ 0x7fff`? Like the intent is to mask off then or right?

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "Re: [miekg/dns] OPT: add Z and SetZ..." ]
>+// SetZ sets the Z part of the OPT RR, note only the 15 least significant bits of z are used.
>+func (rr *OPT) SetZ(z uint16) {
>+       rr.Hdr.Ttl = rr.Hdr.Ttl&0xFFFF7FFF | uint32(z)
>
>Shouldn’t this just be &^ 0x7fff? Like the intent is to mask off then or right?

so:

0x7fff is "0111 1111  1111 1111", with &^ this becomes "1000 0000  0000 0000", so that
leaves the DO bit, before we OR our value back in.

So yes.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
PTAL

File: edns.go [hunk 2]
```diff
@@ -148,6 +148,16 @@ func (rr *OPT) SetDo(do ...bool) {
...
 }
 
+// Z returns the Z part of the OPT RR as a uint16 with only the 15 least significant bits used.
+func (rr *OPT) Z() uint16 {
+	return uint16(rr.Hdr.Ttl & 0x007FFF)
+}
+
+// SetZ sets the Z part of the OPT RR, note only the 15 least significant bits of z are used.
+func (rr *OPT) SetZ(z uint16) {
+	rr.Hdr.Ttl = rr.Hdr.Ttl&^0x7FFF | uint32(z)
```
Comment (tmthrgd) [inline edns.go:158]: The `z` needs a mask as well to match the doc.

File: edns.go [hunk 3]
```diff
@@ -148,6 +148,16 @@ func (rr *OPT) SetDo(do ...bool) {
 	}
 }
 
+// Z returns the Z part of the OPT RR as a uint16 with only the 15 least significant bits used.
+func (rr *OPT) Z() uint16 {
+	return uint16(rr.Hdr.Ttl & 0x007FFF)
```
Comment (tmthrgd) [inline edns.go:153]: Nit: write the mask value the same way in both functions. Either leading zeroes or no leading series.

**Review by tmthrgd — CHANGES_REQUESTED** [review]
Small changes.

File: edns.go (hunk 2 above)
Comment (miekg) [reply edns.go]: and done

**Review by tmthrgd — APPROVED** [review]
LGTM.

**Pull request merged**